Lib/ldap/syncrepl.py

```diff
@@ -535,3 +536,71 @@ def syncrepl_refreshdone(self):
         follows.
         """
         pass
+
+
+class OpenLDAPSyncreplCookie:
+    """
+    OpenLDAPSyncreplCookie - allows a consumer to track a cookie across a
+    refreshAndPersist syncrepl session against a multi-provider OpenLDAP cluster
+    """
+
+    rid: int = 0
+    sid: int = 0
+    _csnset: Dict[int, str]
+
+    def __init__(self, cookie: AnyStr = "") -> None:
+        self._csnset = {}
+
+        if cookie:
+            self.update(cookie)
+
+    def _parse_csn(self, csn: str) -> List[str]:
+        return csn.split('#', 3) # time, order, sid, other
+
+    def _parse_cookie(self, cookie: AnyStr) -> Dict[str, Union[str, List[str]]]:
+        if isinstance(cookie, bytes):
+            cookie = cookie.decode()
+
+        result = {}
+        parts = cookie.split(',')
+        for part in parts:
+            if part.startswith('rid='):
+                result['rid'] = part[4:]
+            elif part.startswith('sid='):
+                result['sid'] = part[4:]
+            elif part.startswith('csn='):
+                result['csn'] = part[4:].split(';')
+            elif part.startswith('delcsn='):
+                result['delcsn'] = part[7:]
+            else:
+                # Did not recognise this cookie part
+                pass
+        return result
+
+    def update(self, cookie: str):
+        """
+        Update the CSN set based on a cookie we just received, use in
+        syncrepl_set_cookie() to track the session state.
+        """
+        components = self._parse_cookie(cookie)
```

Comment:
What about `rid` and `sid` that we parsed (if any)? Should we also assign it here or do I miss something?

Reply:
With OpenLDAP's MPR, signalling a non-zero `sid` carries meaning and puts more responsibility on the consumer. There is no need to track the rid either with the code in python-ldap as-is, so it's zeroed out as well - doing so makes it possible to normalise the cookies if comparison (as a vector clock) is desired.